Collect all error messages for Error.Error()

jegorbunov opened the pull request:
Sometimes mssql server reports a chain of errors with the last error having a message like: `Could not drop constraint. See previous errors.`. Since users code will generally print an error by simply using `.Error()` method to get the details about an underlying error then the method should provide as much detail as possible in my opinion.

## CHANGELOG

**BREAKING CHANGE:**

Before this patch an `Error()` method would return a message like:

`mssql: message`

After this commit it returns:

`mssql: message (error number)`

And even:

```
mssql: message 1 (error number 1)
mssql: message 2 (error number 2)
...
```

jegorbunov committed:
Collect all error messages for Error.Error()

Also error numbers were added to the Error() returned message.

BREAKING CHANGE:

Before this commit an error Error() returned text would look like:

`mssql: message`

After this commit:

`mssql: message (error number)`

And even:

```
mssql: message 1 (error number 1)
mssql: message 2 (error number 2)
...
```

jegorbunov committed:
Collect all error messages for Error.Error()

shueybubbles commented on error.go:
```diff
@@ -24,7 +25,14 @@ type Error struct {
+		msg.WriteString(fmt.Sprintf("mssql: %s (%d)", err.Message, err.Number))
+	}
+	return msg.String()
 }
```
how is this change not affecting any tests? Is there really no coverage of this call?

shueybubbles reviewed (changes requested):
:clock1:

shueybubbles commented on error.go:
```diff
@@ -24,7 +25,14 @@ type Error struct {
 func (e Error) Error() string {
-	return "mssql: " + e.Message
+	var msg strings.Builder
+	for i, err := range e.All {
```
your example implies there are `Error` objects without a Number. Where will `See previous error` show up?

jegorbunov committed:
update TestError/TestMultipleErrors to check Error() result

Before this commit all Error-related tests would not specify
the expected behaviour of Error.Error() method. This commit
adds such expectations so that it is visible when these
expectations change.

jegorbunov committed:
update TestError/TestMultipleErrors to check Error() result

Before this commit all Error-related tests would not specify
the expected behaviour of Error.Error() method. This commit
adds such expectations so that it is visible when these
expectations change.

jegorbunov committed:
fix TestError_Methods test

Error.All must always contain error itself: this is
the requirement which is also checked by TestError
and TestMultipleErrors

dlevy-msft-sql commented on error.go:
```diff
@@ -24,7 +25,14 @@ type Error struct {
 
 func (e Error) Error() string {
-	return "mssql: " + e.Message
+	var msg strings.Builder
```
**Bug:** `Error()` returns `""` when `All` is nil. This breaks any `Error` constructed without `All`, including `token.go:135` (`Error{Message: "Request failed but didn't provide reason"}`) and any external callers.

```suggestion
func (e Error) Error() string {
	if len(e.All) == 0 {
		return "mssql: " + e.Message
	}
	var msg strings.Builder
```

dlevy-msft-sql commented on error_test.go:
```diff
@@ -99,6 +99,9 @@ func TestError_Methods(t *testing.T) {
 		ProcName:   "TestProc",
 		LineNo:     42,
 	}
+	// Error.All must always contain the error itself.
```
This patches the test to work around `Error()` returning `""` when `All` is nil. If the method has a nil fallback (see comment on `error.go`), this isn't needed and the test still passes with the original construction.

dlevy-msft-sql commented on error.go:
```diff
@@ -24,7 +25,14 @@ type Error struct {
+		if i > 0 {
+			msg.WriteByte('\n')
+		}
+		msg.WriteString(fmt.Sprintf("mssql: %s (%d)", err.Message, err.Number))
```
Two design considerations:

1. The format changes from `"mssql: msg"` to `"mssql: msg (num)"` even for single errors -- this is a breaking change for anyone matching on error strings.
2. Newlines in error strings are uncommon in Go. Callers wrapping with `fmt.Errorf("failed: %w", err)` get embedded newlines. Consider `"; "` as a separator or document the choice.
